File: py_functions/mesh_alignment_algorithm.py
import numpy as np
import scipy.io
import scipy.signal
from scipy.stats import find_peaks
from .get_local_coordinate_system import get_local_coordinate_system
from .get_original_pt_location import get_original_pt_location

def mesh_alignment_algorithm(all_verts_time, all_joints_time, person_box_time, person_mask_time, 
                           cams_time, im_shape, cls, cls_category, vid):
    """
    Implements the mesh alignment algorithm via eigen analysis as in sec 3.2 (or 4.2.2) of the paper.
    
    Args:
        all_verts_time (np.ndarray): Coordinates of all mesh points with time before alignment
        all_joints_time (np.ndarray): Coordinates of all joints with time before alignment
        person_box_time (np.ndarray): The bounding box of the person in the video frames
        person_mask_time (list): The binary mask of the person in the video frames
        cams_time (np.ndarray): Camera parameter in all frames
        im_shape (tuple): Dimensions of the video frames in pixels
        cls (str): The activity to be adjusted
        cls_category (int): The category of the activity
        vid (str): Which video of the activity
        
    Returns:
        tuple: (verts, joints, period_start_frames)
            - verts: Coordinates of all mesh points with time after alignment
            - joints: Coordinates of all joints with time after alignment
            - period_start_frames: The frames that indicate the start of each period
    """
    # Load the indices of different body parts in the meshes
    # These should be loaded from .mat files in the MeshInfoMatFiles directory
    try:
        torso_vts_idx = scipy.io.loadmat('MeshInfoMatFiles/torso_vts_idx.mat')['torso_vts_idx'].flatten() - 1
        left_toes_idx = scipy.io.loadmat('MeshInfoMatFiles/left_toes_idx.mat')['left_toes_idx'].flatten() - 1
        right_toes_idx = scipy.io.loadmat('MeshInfoMatFiles/right_toes_idx.mat')['right_toes_idx'].flatten() - 1
        right_fingers_idx = scipy.io.loadmat('MeshInfoMatFiles/right_fingers_idx.mat')['right_fingers_idx'].flatten() - 1
        left_fingers_idx = scipy.io.loadmat('MeshInfoMatFiles/left_fingers_idx.mat')['left_fingers_idx'].flatten() - 1
        leg1_idx = scipy.io.loadmat('MeshInfoMatFiles/leg1_idx.mat')['leg1_idx'].flatten() - 1
        leg2_idx = scipy.io.loadmat('MeshInfoMatFiles/leg2_idx.mat')['leg2_idx'].flatten() - 1
        left_foot_idx = scipy.io.loadmat('MeshInfoMatFiles/left_foot_idx.mat')['left_foot_idx'].flatten() - 1
        right_foot_idx = scipy.io.loadmat('MeshInfoMatFiles/right_foot_idx.mat')['right_foot_idx'].flatten() - 1
        left_thigh_idx = scipy.io.loadmat('MeshInfoMatFiles/left_thigh_idx.mat')['left_thigh_idx'].flatten() - 1
        right_thigh_idx = scipy.io.loadmat('MeshInfoMatFiles/right_thigh_idx.mat')['right_thigh_idx'].flatten() - 1
        hip_idx = scipy.io.loadmat('MeshInfoMatFiles/hip_idx.mat')['hip_idx'].flatten() - 1
    except Exception as e:
        print(f"Error loading body part indices: {e}")
        return all_verts_time, all_joints_time, np.array([])

    front_torso_idx = 3050  # 0-based index (3051-1)
    back_torso_idx = 460    # 0-based index (461-1)

    num_frames = all_verts_time.shape[2]  # total number of frames in video

    # Standing actions w/ one foot/both feet static
    if cls_category in [1, 2]:
        verts = all_verts_time.copy()
        joints = all_joints_time.copy()
        
        # Getting the reference frame (the frame in which the person is standing)
        person_box_height = person_box_time[2,:] - person_box_time[0,:]
        person_box_width = person_box_time[3,:] - person_box_time[1,:]
        
        if cls_category == 1:
            standing_frame = np.argmax(person_box_height)
            # Compute autocorrelation
            person_box_ACF = np.correlate(person_box_height, person_box_height, mode='full')
            person_box_ACF = person_box_ACF[len(person_box_height)-1:]
            person_box_ACF = person_box_ACF[:len(person_box_height)-50]
            peak_locs, _ = find_peaks(person_box_ACF, distance=25)
            action_period = np.mean(np.diff(np.concatenate([[0], peak_locs])))
        else:
            standing_frame = np.argmax(person_box_height - person_box_width)
            try:
                # Compute autocorrelation
                person_box_ACF = np.correlate(person_box_height - person_box_width, 
                                           person_box_height - person_box_width, mode='full')
                person_box_ACF = person_box_ACF[len(person_box_height)-1:]
                person_box_ACF = person_box_ACF[:len(person_box_height)-50]
            except:  # very short video
                person_box_ACF = np.correlate(person_box_height - person_box_width,
                                           person_box_height - person_box_width, mode='full')
                person_box_ACF = person_box_ACF[len(person_box_height)-1:]
                person_box_ACF = person_box_ACF[:len(person_box_height)-1]
            
            peak_locs, _ = find_peaks(person_box_ACF, distance=25)
            action_period = np.mean(np.diff(np.concatenate([[0], peak_locs])))
        
        if cls == 'lateral lunge':
            action_period = action_period * 2  # since we take right and left lunges to be one period
        
        # Get the mesh of the reference frame
        standing_frame_vts = verts[:,:,standing_frame]
        
        # Get the LCS for the reference frame (via eigen analysis)
        localX, localY, localZ = get_local_coordinate_system(
            standing_frame_vts,
            np.concatenate([torso_vts_idx, leg1_idx, leg2_idx]),
            front_torso_idx,
            back_torso_idx,
            left_toes_idx,
            show_plot=False
        )
        
        # Get ankle positions in original frame
        left_ankle_original_standing = get_original_pt_location(
            joints[0,:,standing_frame],
            person_box_time[:,standing_frame],
            cams_time[:,standing_frame],
            im_shape
        )
        right_ankle_original_standing = get_original_pt_location(
            joints[5,:,standing_frame],
            person_box_time[:,standing_frame],
            cams_time[:,standing_frame],
            im_shape
        )
        
        # Test by aligning the standing person's mesh and getting where the feet are in the GCS
        standing_frame_vts_adjusted = (standing_frame_vts - np.mean(standing_frame_vts, axis=0)) @ np.column_stack([localX, localY, localZ])
        left_ankle_fixed = np.mean(standing_frame_vts_adjusted[left_foot_idx,:], axis=0) - \
                          np.mean(standing_frame_vts_adjusted[np.concatenate([left_foot_idx, right_foot_idx]),:], axis=0)
        right_ankle_fixed = np.mean(standing_frame_vts_adjusted[right_foot_idx,:], axis=0) - \
                           np.mean(standing_frame_vts_adjusted[np.concatenate([left_foot_idx, right_foot_idx]),:], axis=0)
        
        for i in range(num_frames):
            # Rotate/align the frame
            verts[:,:,i] = (verts[:,:,i] - np.mean(verts[:,:,i], axis=0)) @ np.column_stack([localX, localY, localZ])
            
            if cls == 'lateral lunge':
                # For lateral lunges, make sure the person does not bend forward in the alignment process
                U, _, _ = np.linalg.svd(verts[:,:,i].T, full_matrices=False)
                which_col = np.argmax(np.abs(U[0,:]))
                U = U * np.sign(U[0,which_col])
                theta = max(np.degrees(np.arctan2(U[2,which_col], U[0,which_col])), 0)
                
                rotation_matrix = np.array([
                    [np.cos(np.radians(theta)), 0, np.sin(np.radians(theta))],
                    [0, 1, 0],
                    [-np.sin(np.radians(theta)), 0, np.cos(np.radians(theta))]
                ])
                verts[:,:,i] = verts[:,:,i] @ rotation_matrix.T
                
                left_foot_check = np.linalg.norm(
                    left_ankle_original_standing - 
                    get_original_pt_location(joints[0,:,i], person_box_time[:,i], cams_time[:,i], im_shape)
                )
                right_foot_check = np.linalg.norm(
                    right_ankle_original_standing - 
                    get_original_pt_location(joints[5,:,i], person_box_time[:,i], cams_time[:,i], im_shape)
                )
                
                if left_foot_check >= right_foot_check:
                    verts[:,:,i] = verts[:,:,i] - np.mean(verts[left_foot_idx,:,i], axis=0) + left_ankle_fixed
                else:
                    verts[:,:,i] = verts[:,:,i] - np.mean(verts[right_foot_idx,:,i], axis=0) + right_ankle_fixed
            
            if cls == 'stiff-leg deadlift':
                # For stiff-leg deadlift, make sure the legs stay up-right
                _, _, localZcurr = get_local_coordinate_system(
                    verts[:,:,i],
                    np.concatenate([leg1_idx, leg2_idx]),
                    front_torso_idx,
                    back_torso_idx,
                    left_toes_idx,
                    show_plot=False
                )
                theta = np.degrees(np.arccos(np.abs(localZcurr[2])/np.linalg.norm([localZcurr[0], localZcurr[2]]))) * 0.75
                
                rotation_matrix = np.array([
                    [np.cos(np.radians(theta)), 0, np.sin(np.radians(theta))],
                    [0, 1, 0],
                    [-np.sin(np.radians(theta)), 0, np.cos(np.radians(theta))]
                ])
                verts[:,:,i] = verts[:,:,i] @ rotation_matrix.T
                verts[:,:,i] = verts[:,:,i] - np.mean(verts[np.concatenate([left_foot_idx, right_foot_idx]),:,i], axis=0)
        
        # Calculate period start frames
        period_start_frames = np.arange(-action_period, num_frames + action_period + 1, action_period)
        frameshift = np.min(np.abs(standing_frame - period_start_frames))
        frameshift_loc = np.argmin(np.abs(standing_frame - period_start_frames))
        period_start_frames = period_start_frames + frameshift * np.sign(standing_frame - period_start_frames[frameshift_loc])
        period_start_frames = period_start_frames[(period_start_frames >= 0) & (period_start_frames < num_frames)]
    
    # Sit up
    elif cls_category == 8:
        verts = all_verts_time.copy()
        joints = all_joints_time.copy()
        
        # Get the reference frame as the widest frame (person lying down)
        person_box_width = person_box_time[3,:] - person_box_time[1,:]
        lying_frame = np.argmax(person_box_width)
        
        lying_frame_vts = verts[:,:,lying_frame]
        
        # Get the LCS
        localX, localY, localZ = get_local_coordinate_system(
            lying_frame_vts,
            np.concatenate([torso_vts_idx, right_foot_idx, left_foot_idx]),
            front_torso_idx,
            back_torso_idx,
            left_toes_idx,
            show_plot=False
        )
        
        # Transform all frames (note now the local X axis points to global Z)
        for i in range(num_frames):
            verts[:,:,i] = (verts[:,:,i] - np.mean(verts[:,:,i], axis=0)) @ np.column_stack([-localZ, localY, localX])
            verts[:,:,i] = verts[:,:,i] - np.mean(verts[hip_idx,:,i], axis=0)
        
        # Compute autocorrelation and find peaks
        person_box_ACF = np.correlate(person_box_width, person_box_width, mode='full')
        person_box_ACF = person_box_ACF[len(person_box_width)-1:]
        person_box_ACF = person_box_ACF[:len(person_box_width)-50]
        peak_locs, _ = find_peaks(person_box_ACF, distance=25)
        action_period = np.mean(np.diff(np.concatenate([[0], peak_locs])))
        
        # Calculate period start frames
        period_start_frames = np.arange(-action_period, num_frames + action_period + 1, action_period)
        frameshift = np.min(np.abs(lying_frame - period_start_frames))
        frameshift_loc = np.argmin(np.abs(lying_frame - period_start_frames))
        period_start_frames = period_start_frames + frameshift * np.sign(lying_frame - period_start_frames[frameshift_loc])
        period_start_frames = period_start_frames[(period_start_frames >= 0) & (period_start_frames < num_frames)]
    
    else:  # unidentified category
        verts = all_verts_time.copy()
        joints = all_joints_time.copy()
        period_start_frames = np.array([])
    
    return verts, joints, period_start_frames 

File: Functions/mesh_alignment_algorithm.m
function [verts, joints, period_start_frames] = mesh_alignment_algorithm(all_verts_time, all_joints_time,...
    person_box_time, person_mask_time, cams_time, im_shape,...
    cls, cls_category, vid)
% MESH ALIGNMENT ALGORITHM  implements the mesh alignment algorithm via
% eigen analysis as in sec 3.2 (or 4.2.2) of the paper.
%   INPUTS:
%	all_verts_time = coordinates of all mesh points with time before alignment
%   all_joints_time = coordinates of all joints with time before alignment
%   person_box_time = the bounding box of the person in the video frames
%   person_mask_time = the binary mask of the person in the video frames
%   cams_time = camera parameter in all frames
%   im_shape = dimensions of the video frames in pixels
%   cls = the activity to be adjusted
%   cls_category = the category of the activity (some activities require
%       same alignment so they get the same category number).
%   vid = which video of the activity
%   OUTPUTS:
%   verts = coordinates of all mesh points with time after alignment
%   joints = coordinates of all joints with time after alignment
%   period_start_frames = the frames that indicate the start of each period
%       in the periodic activities (NaN for non-periodic activities)
% COPYRIGHT (c): H. Cai, B. Korany, and C. Karanam (UCSB, 2020)


% load the indeces of different body parts in the meshes
load torso_vts_idx
load left_toes_idx
load right_toes_idx
load right_fingers_idx
load left_fingers_idx
load leg1_idx
load leg2_idx
load left_foot_idx
load right_foot_idx
load('left_thigh_idx.mat');
load('right_thigh_idx.mat');
load('hip_idx.mat')
front_torso_idx = 3051;
back_torso_idx = 461;

% joints
% 1. ankle left
% 2. knee left
% 3. hip left
% 4. hip right
% 5. knee right
% 6. ankle right
% 7. wrist left
% 8. elbow left
% 9. shoulder left
% 10. shoulder right
% 11. elbow right
% 12. wrist right
% 13. chest/neck
% 14. forehead
% 15. nose
% 16. left eye
% 17. right eye
% 18. left ear
% 19. right ear



num_frames = size(all_verts_time,3); %total number of frames in video

%% ========================================================================
%% standing actions w/ one foot/both feet static, applies to forward lunge,
%% front leg raise, lateral lunge, and stiff-leg deadlift
%% ========================================================================
if cls_category == 1 || cls_category == 2 %|| cls_category == 3
    verts=all_verts_time; 
    joints = all_joints_time; 
    
    % Getting the reference frame (the frame in which the person is
    % standing
    person_box_height = person_box_time(3,:)-person_box_time(1,:);
    person_box_width = person_box_time(4,:)-person_box_time(2,:);
    if cls_category == 1
        [~, standing_frame] = max(person_box_height);
        
        series_cat1 = person_box_height;
        N_cat1 = length(series_cat1);
        disp(['DEBUG: cls_category=1: series length = ', num2str(N_cat1)]);
        person_box_ACF = []; % Initialize
        if N_cat1 >= 2 % Must have at least 2 points for any ACF
            lags_val_cat1 = N_cat1 - 50; % Desired
            disp(['DEBUG: cls_category=1: desired lags (-50) = ', num2str(lags_val_cat1)]);
            if lags_val_cat1 > 0 && lags_val_cat1 < N_cat1 
                person_box_ACF = autocorr(series_cat1, 'NumLags', lags_val_cat1);
                disp('DEBUG: cls_category=1: autocorr with -50 lags SUCCEEDED.');
            else 
                lags_fallback_cat1 = N_cat1 - 1;
                disp(['DEBUG: cls_category=1: -50 lags invalid, trying fallback N-1 lags = ', num2str(lags_fallback_cat1)]);
                if lags_fallback_cat1 > 0 % Ensure N-1 is at least 1
                    person_box_ACF = autocorr(series_cat1, 'NumLags', lags_fallback_cat1);
                    disp('DEBUG: cls_category=1: autocorr with N-1 lags SUCCEEDED.');
                else
                    disp('DEBUG: cls_category=1: N-1 lags also invalid (<=0). ACF set to [].');
                end
            end
        else
            disp('DEBUG: cls_category=1: Series too short for ACF. ACF set to [].');
        end
        if ~isempty(person_box_ACF)
             disp(['DEBUG: cls_category=1: Final ACF class: ', class(person_box_ACF), ', size: [', num2str(size(person_box_ACF,1)), ' ', num2str(size(person_box_ACF,2)), ']']);
        end

        min_peak_dist = floor(length(person_box_ACF) / 2) - 1;
        if min_peak_dist < 1
            min_peak_dist = 1;
        end
        [~,peak_locs] = findpeaks(person_box_ACF,'MinPeakDistance',min_peak_dist);
        
        if isempty(peak_locs)
            action_period = NaN; % No peaks found, so period is undefined or use a default
            disp('DEBUG: No peaks found by findpeaks. action_period set to NaN.');
        else
            peak_locs = peak_locs(:)'; % Ensure peak_locs is a row vector for [0 peak_locs]
            action_period = mean(diff([0 peak_locs]));
            disp(['DEBUG: Peaks found. action_period calculated as: ', num2str(action_period)]);
        end
    else
        [~, standing_frame] = max(person_box_height- person_box_width);

        debug_series_data = person_box_height - person_box_width;
        disp('DEBUG: About to call autocorr (cls_category not 1). Series data properties:');
        disp(['  Class: ', class(debug_series_data)]);
        disp(['  Size: [', num2str(size(debug_series_data,1)), ' ', num2str(size(debug_series_data,2)), ']']);
        disp(['  Is all finite? ', num2str(all(isfinite(debug_series_data)))]');
        disp(['  Is real? ', num2str(isreal(debug_series_data))]);
        
        person_box_ACF = []; % Initialize
        try
            disp('DEBUG: Trying autocorr in TRY block for cls_category not 1');
            raw_series_try = person_box_height- person_box_width; % Same as debug_series_data
            N_try = length(raw_series_try);
            lags_try = N_try - 50;
            disp(['DEBUG: TRY block: series length = ', num2str(N_try), ', lags (-50) = ', num2str(lags_try)]);

            if N_try >=2 
                if lags_try > 0 && lags_try < N_try
                    person_box_ACF = autocorr(raw_series_try, 'NumLags', lags_try);
                    disp('DEBUG: autocorr in TRY succeeded. Class of person_box_ACF:'); disp(class(person_box_ACF)); disp('Size:'); disp(size(person_box_ACF));
                else 
                    disp('DEBUG: TRY block: Invalid lags calculated (-50), forcing to CATCH for N-1 lags.');
                    error('Calculated lags for -50 are invalid, using N-1 via catch block.');
                end
            else 
                 disp('DEBUG: TRY block: Series too short for autocorr (N<2), forcing to CATCH.');
                 error('Series too short in TRY (N<2), using N-1 via catch block.');
            end
        catch ME_autocorr_try 
            disp(['DEBUG: CAUGHT error/condition in TRY block autocorr. Message: ', ME_autocorr_try.message]);
            disp('DEBUG: Trying autocorr in CATCH block (using N-1 lags if possible)');
            temp_series_catch = person_box_height - person_box_width; 
            N_catch = length(temp_series_catch);
            disp(['DEBUG: CATCH block: series length = ', num2str(N_catch)]);

            if N_catch >= 2 
                lags_to_use_catch = N_catch - 1;
                disp(['DEBUG: CATCH block: lags_to_use = ', num2str(lags_to_use_catch)]);
                person_box_ACF = autocorr(temp_series_catch, 'NumLags', lags_to_use_catch);
                disp('DEBUG: autocorr in CATCH succeeded. Class of person_box_ACF:'); disp(class(person_box_ACF)); disp('Size:'); disp(size(person_box_ACF));
            else
                disp('DEBUG: CATCH block: Series too short for autocorr (N<2).');
                % person_box_ACF is already [], or will remain [] if try was skipped by error above
                disp('DEBUG: CATCH block: person_box_ACF set/kept as []. Class:'); disp(class(person_box_ACF)); disp('Size:'); disp(size(person_box_ACF));
            end
        end
        
        % Defensive checks before findpeaks
        if ~isnumeric(person_box_ACF) || (~isvector(person_box_ACF) && ~isempty(person_box_ACF))
            disp('Warning: person_box_ACF is not a numeric vector or is not empty and not a vector. Setting to empty for findpeaks.');
            person_box_ACF = [];
        end
        % Ensure it's a column vector if not empty and is a vector
        if ~isempty(person_box_ACF) && isvector(person_box_ACF) && ~iscolumn(person_box_ACF) 
            person_box_ACF = person_box_ACF(:);
        end
        min_peak_dist = floor(length(person_box_ACF) / 2) - 1;
        if min_peak_dist < 1
            min_peak_dist = 1;
        end
        [~,peak_locs] = findpeaks(person_box_ACF,'MinPeakDistance',min_peak_dist);
        
        if isempty(peak_locs)
            action_period = NaN; % No peaks found, so period is undefined or use a default
            disp('DEBUG: No peaks found by findpeaks. action_period set to NaN.');
        else
            peak_locs = peak_locs(:)'; % Ensure peak_locs is a row vector for [0 peak_locs]
            action_period = mean(diff([0 peak_locs]));
            disp(['DEBUG: Peaks found. action_period calculated as: ', num2str(action_period)]);
        end
    end
    
    if strcmp(cls,'lateral lunge')
        action_period = action_period*2; %since we take right and left lunges to be one period
    end

    
    % get the mesh of the reference frame
    standing_frame_vts = verts(:,:, standing_frame);
    % get the LCS for the reference fram (via eigen analysis)
    [localX, localY, localZ] = get_local_coordinate_system(standing_frame_vts,[torso_vts_idx; leg1_idx; leg2_idx], ...
        front_torso_idx,back_torso_idx, left_toes_idx, 0);
    
    left_ankle_original_standing = get_original_pt_location(joints(1,:,standing_frame), person_box_time(:, standing_frame), cams_time(:,standing_frame), im_shape);
    right_ankle_original_standing = get_original_pt_location(joints(6,:,standing_frame), person_box_time(:, standing_frame), cams_time(:,standing_frame), im_shape);
    
    %test by aligning the standing person's mesh and getting where the feet
    %are in the GCS
    standing_frame_vts_adjusted = (standing_frame_vts - mean(standing_frame_vts,1))*[localX, localY, localZ];
    left_ankle_fixed  = mean(standing_frame_vts_adjusted(left_foot_idx,:)) - mean(standing_frame_vts_adjusted([left_foot_idx; right_foot_idx],:));
    right_ankle_fixed  = mean(standing_frame_vts_adjusted(right_foot_idx,:)) - mean(standing_frame_vts_adjusted([left_foot_idx; right_foot_idx],:));
    
    for i = 1 : num_frames %loop over all frames
        verts(:,:,i) = (verts(:,:,i)-mean(verts(:,:,i),1))*[localX, localY, localZ]; %rotate/align the frame
        
        if strcmp(cls, 'lateral lunge') %for lateral lunges, make sure the person does not bend forward in the alignment process
            [U,~,~] = svd(verts(:,:,i)','econ');
            [~,which_col] = max(abs(U(1,:)));
            U = U * sign(U(1,which_col));
            theta = max(atand(U(3,which_col)/ U(1,which_col)),0);
            verts(:,:,i) = verts(:,:,i)*[cosd(theta),0,sind(theta);0,1,0;-sind(theta),0,cosd(theta)]';
            left_foot_check = norm(left_ankle_original_standing - get_original_pt_location(joints(1,:,i), person_box_time(:, i), cams_time(:,i), im_shape));
            right_foot_check = norm(right_ankle_original_standing - get_original_pt_location(joints(6,:,i), person_box_time(:, i), cams_time(:,i), im_shape));
            if left_foot_check >= right_foot_check                    
                verts(:,:,i) = verts(:,:,i) - mean(verts(left_foot_idx,:,i)) + left_ankle_fixed;
            else
                verts(:,:,i) = verts(:,:,i) - mean(verts(right_foot_idx,:,i)) + right_ankle_fixed;
            end
        end        
        if strcmp(cls, 'stiff-leg deadlift') % for stiff-leg deadlift, make sure the legs stay up-right
            [~, ~, localZcurr] = get_local_coordinate_system(verts(:,:,i),[leg1_idx; leg2_idx], ...
                front_torso_idx,back_torso_idx, left_toes_idx, 0);
            theta = acosd(abs(localZcurr(3))/norm([localZcurr(1) localZcurr(3)]))*0.75;
            verts(:,:,i) = verts(:,:,i)*[cosd(theta),0,sind(theta);0,1,0;-sind(theta),0,cosd(theta)]';
            verts(:,:,i) = verts(:,:,i) - mean(verts([left_foot_idx; right_foot_idx],:,i));
        end
    end
    
    period_start_frames = -action_period:action_period: num_frames+action_period;
    [frameshift, frameshift_loc]  = min(abs(standing_frame- period_start_frames));
    period_start_frames = period_start_frames + frameshift*sign(standing_frame-period_start_frames(frameshift_loc));
    period_start_frames(period_start_frames < 0) = [];
    period_start_frames(period_start_frames > num_frames) = [];
    

%% ========================================================================
%% sit up
%% ========================================================================
elseif cls_category == 8
    verts=all_verts_time;
    joints = all_joints_time;
    
    
    % Get the reference frame as the widest frame (person lying down)
    person_box_width = person_box_time(4,:)-person_box_time(2,:);
    [~, lying_frame] = max(person_box_width);
    
    lying_frame_vts = verts(:,:, lying_frame);
    
    % Get the LCS
    [localX, localY, localZ] = get_local_coordinate_system(lying_frame_vts,[torso_vts_idx;right_foot_idx; left_foot_idx], ...
        front_torso_idx,back_torso_idx, left_toes_idx, 0);
    
    
    % Transform all frame (note now the local X axis points to global Z)
    for i = 1 : num_frames
        verts(:,:,i) = (verts(:,:,i)-mean(verts(:,:,i),1))*[ -localZ , localY ,  localX];
        verts(:,:,i) = verts(:,:,i) - mean(verts(hip_idx,:,i));
    end
    
    % Robust autocorr for person_box_width (cls_category == 8)
    series_cat8 = person_box_width;
    N_cat8 = length(series_cat8);
    person_box_ACF_cat8 = []; % Initialize
    disp(['DEBUG: cls_category=8: series length = ', num2str(N_cat8)]);
    if N_cat8 >= 2 % Must have at least 2 points for any ACF
        lags_desired_cat8 = N_cat8 - 50;
        disp(['DEBUG: cls_category=8: desired lags (-50) = ', num2str(lags_desired_cat8)]);
        if lags_desired_cat8 > 0 && lags_desired_cat8 < N_cat8
            person_box_ACF_cat8 = autocorr(series_cat8, 'NumLags', lags_desired_cat8);
            disp('DEBUG: cls_category=8: autocorr with -50 lags SUCCEEDED.');
        else
            lags_fallback_cat8 = N_cat8 - 1;
            disp(['DEBUG: cls_category=8: -50 lags invalid, trying N-1 lags = ', num2str(lags_fallback_cat8)]);
            if lags_fallback_cat8 > 0 % Ensure N-1 is at least 1
                person_box_ACF_cat8 = autocorr(series_cat8, 'NumLags', lags_fallback_cat8);
                disp('DEBUG: cls_category=8: autocorr with N-1 lags SUCCEEDED.');
            else
                disp('DEBUG: cls_category=8: N-1 lags also invalid (<=0). ACF set to [].');
            end
        end
    else
        disp('DEBUG: cls_category=8: Series too short for ACF. ACF set to [].');
    end
    if ~isempty(person_box_ACF_cat8)
        disp(['DEBUG: cls_category=8: Final ACF class: ', class(person_box_ACF_cat8), ', size: [', num2str(size(person_box_ACF_cat8,1)), ' ', num2str(size(person_box_ACF_cat8,2)), ']']);
    end
    person_box_ACF = person_box_ACF_cat8; % Assign to the variable used by findpeaks

    min_peak_dist = floor(length(person_box_ACF) / 2) - 1;
    if min_peak_dist < 1
        min_peak_dist = 1;
    end
    [~,peak_locs] = findpeaks(person_box_ACF,'MinPeakDistance',min_peak_dist);
    action_period = mean(diff([0 peak_locs]));
    period_start_frames = -action_period:action_period: num_frames+action_period;
    [frameshift, frameshift_loc]  = min(abs(lying_frame - period_start_frames));
    period_start_frames = period_start_frames + frameshift*sign(lying_frame-period_start_frames(frameshift_loc));
    period_start_frames(period_start_frames < 0) = [];
    period_start_frames(period_start_frames > num_frames) = [];
    
    

else %unidentified category
    verts = all_verts_time;
    joints = all_joints_time;
end